forest_fuzz crash testing

sm2n opened the pull request:
Add crash and reset storage fuzz_op so that we can test that forest checkpointing works correctly. This can't test crashes that are concurrent with checkpointing though.

## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

sm2n committed:
LSM: Properly crash stuff in forest_fuzz

sm2n committed:
LSM: forest_fuzz fuzz_op generator generate correct op number

sm2n committed:
LSM/forest_fuzz: Properly generate op numbers

Checkpointed (to disk) data is a bar behind compaction work.

sm2n committed:
Add build options to generate executables for every fuzzer

Easier to open in a debugger this way.

sm2n committed:
LSM: add crash/restart to forest_fuzz

This doesn't simulate crashing during a checkpoint, but it does
interleave crashing with other operations.

sm2n committed:
LSM/forest_fuzz: move fuzzer business logic into separate function

This made it easier to step in a debugger.

sm2n committed:
LSM/forest_fuzz: lower the probability of storage resets

If it's too high, the fuzzer won't make much progress, so we won't
explore interesting behavior

sm2n committed:
LSM/forest_fuzz: don't copy the model tuple around

sm2n committed:
LSM/forest_fuzz: properly model checkpoints

sm2n committed:
LSM/forest_fuzz: Properly model checkpoints, for real this time (mostly)

Gave up on the hashmap waterfall based model, now we just use a queue
and do a linear scan as a model of the (im)mutable tables of the
lsm. It seems to mostly work now.

kprotty commented on src/lsm/forest_fuzz.zig:
```diff
@@ -273,41 +355,81 @@ const Environment = struct {
+                env.storage.reset();
+                try env.init(env.storage);
+
+                env.state = .superblock_open;
```
prefer `env.change_state(.init, .superblock_open)`

sm2n replied:
fixed

kprotty commented on src/lsm/forest_fuzz.zig:
```diff
@@ -273,41 +355,81 @@ const Environment = struct {
+                    if (model.checkpointed.get(id)) |checkpointed_account| {
+                        env.prefetch_account(id);
+                        if (env.get_account(id)) |lsm_account| {
+                            assert(std.mem.eql(
```
Should this also assert that the lsm_account != entry.account?

sm2n replied:
If `lsm_account` == `entry.account` then `lsm_account` != `checkpointed_account`, unless `entry.account` == `checkpointed_account`, in which case it's a false positive anyway. So I don't think it buys anything.

sm2n committed:
LSM/forest_fuzz: Cleaner fuzzer state machine changes

sm2n committed:
LSM/forest_fuzz: small clean up

sm2n commented:
(reposting)
The fuzzer found a crash that exists because the fuzzer doesn't always checkpoint at the end of each bar.
https://github.com/tigerbeetledb/tigerbeetle/blob/a5756448caec336159776d79bb9489af8a0c033e/src/lsm/tree.zig#L974
The intent here seems to be that we don't checkpoint at the same op twice, but the effect is that if we don't always checkpoint, we can:

1. checkpoint on bar 1
2. don't checkpoint on bar 2
3. crash during bar 3
4. recover from bar 1 checkpoint
5. crash trying to checkpoint in bar 2

is this a real bug or should I change the fuzzer generator? I think it should be in this PR if so, because currently the fuzzer crashes very fast, and doesn't get to explore much.

sm2n committed:
LSM/forest_fuzz: Discard prefix of log on checkpoint

Stop leaking memory

sm2n committed:
LSM/forest_fuzz: Always checkpoint when possible

Matches how it works with the real vsr replica code, where we always
checkpoint at the end of a bar.

sm2n committed:
LSM/forest_fuzz: Make the probability of storage_reset more variable

This way we can explore the behavior of long runs without crashes and
also very radioactive machines.

sm2n commented:
I changed how the fuzzop generator works to match how vsr replicas work wrt checkpointing. I still don't know why it was the way it was though, and would like to know.

I ran a single run and it passed, will run fuzz_loop in the background today.

jamii commented:
> don't checkpoint at the same op twice
> 5\. crash trying to checkpoint in bar 2

I don't follow this explanation. We didn't previously checkpoint on bar 2, so why is checkpointing on bar 2 an error?

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -262,8 +277,77 @@ const Environment = struct {
+                const log_size = model.log.readableLength();
+                var log_left = log_size;
+                while (log_left > 0) : (log_left -= 1) {
+                    const entry = model.log.peekItem(log_size - log_left);
```
Since you use `log_size - log_left` everywhere this seems to be equivalent to:

``` zig
var log_index: usize = 0;
while (log_index < log_size) : (log_index += 1) {
    const entry = model.log.peekItem(log_index);
    if (entry.op > checkpointable) break;
    try model.checkpointed.put(entry.account.id, entry.account);
}
model.log.discard(log_index);
```

sm2n replied:
Good catch! Will change

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -273,41 +357,83 @@ const Environment = struct {
 
-            const model_size = model.count() * @sizeOf(Account);
+            const model_size = (model.log.readableLength() + model.checkpointed.count()) * @sizeOf(Account);
+            // FIXME: This isn't accurate anymore because the model can contain multiple copies of an account in the log
```
That's unfortunate, but at least it's asymptotically accurate.

sm2n replied:
I think we'd have to maintain a separate set data structure for this if we wanted to keep it, do you think it's worth the extra overhead?

jamii replied:
I think you can leave it as is. Maybe change the `=` in the log message to `~=` :)

sm2n replied:
good idea

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -273,41 +357,83 @@ const Environment = struct {
+                const log_size = model.log.readableLength();
+                var log_left = log_size;
+                while (log_left > 0) : (log_left -= 1) {
+                    const entry = model.log.peekItem(log_size - log_left);
```
Again, if you only use log_size - log_left, might as well start at 0.

sm2n replied:
will change

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -370,8 +499,11 @@ pub fn generate_fuzz_ops(random: std.rand.Random, fuzz_op_count: usize) ![]const
-                    // Checkpoint at roughly the same rate as log wraparound.
-                    random.uintLessThan(usize, Environment.compacts_per_checkpoint) == 0;
+                    // Never checkpoint at the same op twice
+                    compact_op > checkpointed_op + constants.lsm_batch_multiple;
```
This now checkpoints at the end of every measure. The replica checkpoints (slightly more than) every `journal_slot_count / lsm_batch_multiple == 16` measures. That's what the `compacts_per_checkpoint` was for.

sm2n replied:
Yeah, that was what I intended, but now I understand that that was the wrong thing to do.

sm2n committed:
LSM/forest_fuzz: more natural forward log loops

sm2n commented:
> > don't checkpoint at the same op twice
> > 5. crash trying to checkpoint in bar 2
> 
> I don't follow this explanation. We didn't previously checkpoint on bar 2, so why is checkpointing on bar 2 an error?

Hmm, that's a good question. I don't know how I got the idea, but I was under the impression that compaction could increase `commit_min`, but looking at the code again it seems that only happens on checkpoint... I'll take another look at the crash I found before.

sm2n commented:
Okay, I think I understand now after looking at a bunch of crashes. I made a mistake before, in that I mixed up compaction and checkpoint (compaction + batch size) op numbers. The crash happens when an op is checkpointed twice. This can't happen in the replica code because compaction is written to be idempotent, and checkpoint is called by compaction as a callback, so it's gated. But in the fuzzer we synchronously start checkpointing without going through the compaction codepath.

So, the solution is almost what I had before. I'll update the branch now.

sm2n committed:
LSM/forest_fuzz: Bring back the random checkpointing

The replica doesn't actually checkpoint every bar. However, we do need
to make sure to not checkpoint the same op twice, as in the replica
code it is gated by the compaction code, but that's not the case in
the fuzzer.

sm2n committed:
LSM/forest_fuzz: Update note on the space amplification debug log

sm2n committed:
LSM/forest_fuzz: Clean up loop

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -370,8 +498,13 @@ pub fn generate_fuzz_ops(random: std.rand.Random, fuzz_op_count: usize) ![]const
                     compact_op % constants.lsm_batch_multiple == constants.lsm_batch_multiple - 1 and
                     compact_op > constants.lsm_batch_multiple and
+                    // Never checkpoint at the same op twice
+                    compact_op > checkpointed_op + constants.lsm_batch_multiple and
```
Curious - do we need `compact_op > checkpointed_op + constants.lsm_batch_multiple` or is `compact_op > checkpointed_op` sufficient? 

`compact_op % constants.lsm_batch_multiple == constants.lsm_batch_multiple - 1` allows checkpoints to be one measure apart, but `compact_op > checkpointed_op + constants.lsm_batch_multiple` requires them to be at least two measures apart. Is that intentional?

sm2n replied:
I think it's correct as it is. Currently, `checkpointed_op` is the highest op number that has made it to disk, not the op number when the checkpoint actually happened (which is one bar ahead). So it's actually not two measures apart, only one.

Thinking about it, the naming is clearly confusing. I'll rename it to `persisted_op`.

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -370,8 +498,13 @@ pub fn generate_fuzz_ops(random: std.rand.Random, fuzz_op_count: usize) ![]const
                     // Checkpoint at roughly the same rate as log wraparound.
                     random.uintLessThan(usize, Environment.compacts_per_checkpoint) == 0;
+                if (checkpoint) {
+                    checkpointed_op = op - constants.lsm_batch_multiple;
```
Is this reachable when `op == constants.lsm_batch_multiple - 1`?

jamii replied:
I guess only reachable with the change I suggested above. In which case changing `-` to `-|` might be an easy fix.

sm2n replied:
Even with the change you suggested it's still unreachable because one of the other conjuncts is `compact_op > constants.lsm_batch_multiple`.

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -262,8 +277,76 @@ const Environment = struct {
+            pub fn checkpoint(model: *Model, op: u64) !void {
+                const checkpointable = blk: {
+                    if (op > constants.lsm_batch_multiple) {
+                        break :blk op - (op % constants.lsm_batch_multiple) - 1;
```
This block can just be `const checkpointable = op - (op % constants.lsm_batch_multiple) -| 1`.

sm2n replied:
thanks, changed

sm2n committed:
LSM/forest_fuzz: rename checkpointed_op to persisted_op

sm2n committed:
LSM/forest_fuzz: clean up block with saturating subtraction

sm2n committed:
Merge branch 'main' into sm2n/forest-fuzz-crash

jamii commented:
How long does a fuzzer run usually take for you now? I've been running one for 15 minutes and it hasn't finished yet.

jamii commented:
Looks like open and init are really expensive because of all the allocation. I'll turn the reset rate way down.

sm2n commented:
> How long does a fuzzer run usually take for you now? I've been running one for 15 minutes and it hasn't finished yet.

To be honest, off the top of my head I've had runs that were way longer and also way shorter than that, and I didn't record any information when I was running it in a loop to get an idea of the distribution vs the number of fuzzops.

But the variability in fuzzops is a lot, back before I actually wrote any code, when I was running the fuzzer prior to this pr, some runs took <10 mins and some took >45min. The fuzzer can currently generate up to 10 million ops, which seems like a lot.

jamii committed:
Reduce reset rate for fuzzer runs dont take forever

sm2n commented:
Oh, another thing is that the new Model code can be pathologically worse, since it's doing a linear scan through the log on every get. Normally this shouldn't be too much of an issue since checkpointing truncates the log relatively often, but I suppose it's possible to get unlucky since there's no actual bound.

jamii commented:
I tried fixing `--seed 483660332232550540` which has only 80k ops. With the reset rate set to 0 it takes 6 secs. With the reset rate set to 1 it's taken over 15 minutes so far :grimacing: 

I've dialed the rate down to 1e-4 and verified that that's still high enough to cause a couple of resets on larger runs.

sm2n commented:
Wow yeah that's a big difference. I do think it's worth varying the rate of resets though, since there's an interplay between the time between checkpoints and resets which I think is where there would be interesting behavior when it comes to testing resets.

jamii commented:
The fuzz ops get picked randomly from that distribution, so the number of fuzz ops between checkpoint and reset is still exponentially distributed. We could play with varying the rate more, but I'm wary of setting it to something open-ended and then having some small percentage of fuzz runs not terminate.
